samples/WinForms/Dx12Viewport.cs

```diff
@@ -0,0 +1,769 @@
+// Copyright © Tanner Gooding and Contributors. Licensed under the MIT License (MIT). See License.md in the repository root for more information.
+
+// Ported from D3D12HelloWindow.h in https://github.com/Microsoft/DirectX-Graphics-Samples
+// Original source is Copyright © Microsoft. All rights reserved. Licensed under the MIT License (MIT).
+
+using System;
+using System.IO;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using TerraFX.Interop;
+using static TerraFX.Interop.D3D_FEATURE_LEVEL;
+using static TerraFX.Interop.D3D_PRIMITIVE_TOPOLOGY;
+using static TerraFX.Interop.D3D_ROOT_SIGNATURE_VERSION;
+using static TerraFX.Interop.D3D12;
+using static TerraFX.Interop.D3D12_COMMAND_LIST_TYPE;
+using static TerraFX.Interop.D3D12_DESCRIPTOR_HEAP_TYPE;
+using static TerraFX.Interop.D3D12_FENCE_FLAGS;
+using static TerraFX.Interop.D3D12_HEAP_FLAGS;
+using static TerraFX.Interop.D3D12_HEAP_TYPE;
+using static TerraFX.Interop.D3D12_INPUT_CLASSIFICATION;
+using static TerraFX.Interop.D3D12_PRIMITIVE_TOPOLOGY_TYPE;
+using static TerraFX.Interop.D3D12_RESOURCE_STATES;
+using static TerraFX.Interop.D3D12_ROOT_SIGNATURE_FLAGS;
+using static TerraFX.Interop.D3DCompiler;
+using static TerraFX.Interop.DXGI;
+using static TerraFX.Interop.DXGI_ADAPTER_FLAG;
+using static TerraFX.Interop.DXGI_FORMAT;
+using static TerraFX.Interop.DXGI_SWAP_EFFECT;
+using static TerraFX.Interop.Kernel32;
+using static TerraFX.Interop.Windows;
+using static TerraFX.Samples.DirectX.DXSampleHelper;
+
+namespace TerraFX.Samples.DirectX.D3D12
+{
+    public unsafe class DX12Viewport : IDisposable
+    {
+
+        // Viewport dimensions
+        private uint _width;
+
+        private uint _height;
+
+        private float _aspectRatio;
+
+        // Adapter info
+        private readonly bool _useWarpDevice;
+
+        // Root assets path
+        private readonly string _assetsPath;
+
+        private const uint FrameCount = 2;
+
+        // Pipeline objects
+        private D3D12_VIEWPORT _viewport;
+        private RECT _scissorRect;
+        private IDXGISwapChain3* _swapChain;
+        private ID3D12Device* _device;
+        private readonly ID3D12Resource*[] _renderTargets;
+        private ID3D12CommandAllocator* _commandAllocator;
+        private ID3D12CommandQueue* _commandQueue;
+        private ID3D12RootSignature* _rootSignature;
+        private ID3D12DescriptorHeap* _rtvHeap;
+        private ID3D12PipelineState* _pipelineState;
+        private ID3D12GraphicsCommandList* _commandList;
+        private uint _rtvDescriptorSize;
+        private readonly IntPtr _hWnd;
+        private bool _isOnColorToggle;
```

Comment:
This really shouldn't be part of the viewport itself... Ideally the viewport would just be a generic/reusable and sample specific things would be tracked in the respective sample.

I'm also not sure `Viewport` is a good name here, this is really all the bits needed for the `DX12Panel` and the two classes should probably be merged.
Bits like the `Width`, `Height`, and `_hWnd` would then come from the `Panel` object that `DX12Panel` inherits from.

Reply:
Makes sense.  Can I do that in a subsequent PR?

Reply:
actually, ignore the above comment. Refactoring now. Will be a separate commit. That should be enough to encapsulate this change. Does not need a new PR.

Reply:
done.